x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/latest/LatestDoc.java

```diff
@@ -0,0 +1,275 @@
+import org.elasticsearch.xpack.core.transform.transforms.TransformProgress;
+import org.elasticsearch.xpack.core.transform.transforms.latest.LatestDocConfig;
+import org.elasticsearch.xpack.transform.transforms.Function;
+import org.elasticsearch.xpack.transform.transforms.IDGenerator;
+import org.elasticsearch.xpack.transform.transforms.common.DocumentConversionUtils;
+import org.elasticsearch.xpack.transform.transforms.pivot.SchemaUtil;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static java.util.function.Predicate.not;
+import static java.util.stream.Collectors.toMap;
+import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
+
+public class LatestDoc implements Function {
+
+    public static final int DEFAULT_INITIAL_MAX_PAGE_SEARCH_SIZE = 5000;
+    public static final int TEST_QUERY_PAGE_SIZE = 50;
+
+    private static final String COMPOSITE_AGGREGATION_NAME = "_transform";
+    private static final String TOP_HITS_AGGREGATION_NAME = "_top_hits";
+    private static final Logger logger = LogManager.getLogger(LatestDoc.class);
+
+    private final LatestDocConfig config;
+
+    // objects for re-using
+    private final CompositeAggregationBuilder cachedCompositeAggregation;
+
+    public LatestDoc(LatestDocConfig config) {
+        this.config = config;
+        this.cachedCompositeAggregation = createCompositeAggregation(config);
+    }
+
+    private static CompositeAggregationBuilder createCompositeAggregation(LatestDocConfig config) {
+        CompositeAggregationBuilder compositeAggregation;
+        try (XContentBuilder builder = jsonBuilder()) {
+            config.toCompositeAggXContent(builder);
+            XContentParser parser = builder.generator()
+                .contentType()
+                .xContent()
+                .createParser(NamedXContentRegistry.EMPTY, LoggingDeprecationHandler.INSTANCE, BytesReference.bytes(builder).streamInput());
+            compositeAggregation = CompositeAggregationBuilder.PARSER.parse(parser, COMPOSITE_AGGREGATION_NAME);
+        } catch (IOException e) {
+            throw new RuntimeException(
+                TransformMessages.getMessage(TransformMessages.TRANSFORM_FAILED_TO_CREATE_COMPOSITE_AGGREGATION, "latest"), e);
+        }
+        TopHitsAggregationBuilder topHitsAgg =
+            AggregationBuilders.topHits(TOP_HITS_AGGREGATION_NAME)
+                .size(1)  // we are only interested in the top-1
+                .sorts(config.getSorts());  // we copy the sort config directly from the function config
+        compositeAggregation.subAggregation(topHitsAgg);
+        return compositeAggregation;
+    }
+
+    @Override
+    public int getInitialPageSize() {
+        return DEFAULT_INITIAL_MAX_PAGE_SEARCH_SIZE;
+    }
+
+    private SearchRequest buildSearchRequest(SourceConfig sourceConfig, Map<String, Object> position, int pageSize) {
```

Comment:
you might get rid of this method or at least rename it. Another example of legacy in pivot, better not to copy (in pivot we ended up with this after several re-factorings, it should be fixed there, too!).

The name is misleading, this is called for validation and preview, maybe you can name it accordingly. Or if you think re-usage makes no sense inline it.